voxblox/include/voxblox/utils/neighbor_tools.h

```diff
@@ -13,246 +13,87 @@ enum Connectivity : unsigned int {
+
+    for (unsigned int i = 0u; i < 3u; ++i) {
+      while ((*neighbor_voxel_index)(i) < 0) {
+        (*neighbor_block_index)(i)--;
+        (*neighbor_voxel_index)(i) += voxels_per_side;
       }
-      direction(i) = 0;
-      direction(next_i) = 0;
-    }
-  }
-
-  if (connectivity == Connectivity::kEighteen) {
-    return;
-  }
-
-  // Distance sqrt(3) set.
-  for (int i = -1; i <= 1; i += 2) {
-    direction(0) = i;
-    for (int j = -1; j <= 1; j += 2) {
-      direction(1) = j;
-      for (int k = -1; k <= 1; k += 2) {
-        direction(2) = k;
-        neighbors->emplace_back(global_index + direction);
+      while ((*neighbor_voxel_index)(i) >=
+             static_cast<IndexElement>(voxels_per_side)) {
+        (*neighbor_block_index)(i)++;
+        (*neighbor_voxel_index)(i) -= voxels_per_side;
       }
     }
   }
-  if (connectivity != Connectivity::kTwentySix) {
-    LOG(FATAL) << "Unknown neighborhood connectivity: "
-               << static_cast<int>(connectivity);
-  }
-}
-
-template <typename VoxelType>
-void NeighborTools<VoxelType>::getNeighborIndex(
-    const BlockIndex& block_index, const VoxelIndex& voxel_index,
-    const SignedIndex& direction, BlockIndex* neighbor_block_index,
-    VoxelIndex* neighbor_voxel_index) const {
-  CHECK_NE(voxels_per_side_, 0u);
-  CHECK_NOTNULL(neighbor_block_index);
-  CHECK_NOTNULL(neighbor_voxel_index);
-
-  *neighbor_block_index = block_index;
-  *neighbor_voxel_index = voxel_index + direction;
 
-  for (unsigned int i = 0; i < 3; ++i) {
-    while ((*neighbor_voxel_index)(i) < 0) {
-      (*neighbor_block_index)(i)--;
-      (*neighbor_voxel_index)(i) += voxels_per_side_;
-    }
-    while ((*neighbor_voxel_index)(i) >=
-           static_cast<IndexElement>(voxels_per_side_)) {
-      (*neighbor_block_index)(i)++;
-      (*neighbor_voxel_index)(i) -= voxels_per_side_;
+  // Get the hierarchical indices (block idx, local voxel index) for all
+  // neighbors (6, 18, or 26 neighborhood) of a hierarcical index. This function
+  // solves the cross-block indexing that happens when looking up neighbors at
+  // the block boundary.
+  static void getFromHierarchicalIndex(const BlockIndex& block_index,
+                                       const VoxelIndex& voxel_index,
+                                       const size_t voxels_per_side,
+                                       AlignedVector<VoxelKey>* neighbors_ptr) {
+    CHECK_NOTNULL(neighbors_ptr)->resize(kConnectivity);
+
+    AlignedVector<VoxelKey>& neighbors = *neighbors_ptr;
+    for (unsigned int i = 0u; i < kConnectivity; ++i) {
+      VoxelKey& neighbor = neighbors[i];
+      getFromHierarchicalIndexAndDirection(block_index, voxel_index,
+                                           kOffsets.col(i), voxels_per_side,
+                                           &neighbor.first, &neighbor.second);
     }
   }
-}
-
-template <typename VoxelType>
-SignedIndex NeighborTools<VoxelType>::getOffsetBetweenVoxels(
```

Comment:
I use this in the skeleton generation and planning. Please don't remove this. :( :( :( :( : (